File: server_354/src/main/java/com/example/server_354/controller/VehicleController.java
package com.example.server_354.controller;

import com.example.server_354.Services.VehicleService;
import com.example.server_354.object.*;
import org.springframework.dao.DataAccessException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.*;

@RestController
@RequestMapping("/vehicle")
public class VehicleController {

    private final VehicleService vehicleService;

    public VehicleController(VehicleService vehicleService){
        this.vehicleService = vehicleService;
    }

    public String substringIncomingString(String text){
        return text.substring(1, text.length() - 1);
    }

    @GetMapping("/getMake")
    public ResponseEntity<?> getVehicleMake() {
        Map<String, String> errorResponse = new HashMap<>();
        try {
            return ResponseEntity.ok(vehicleService.getEveryMake());
        } catch (DataAccessException e) {
            System.out.println(e.getMessage());
            errorResponse.put("error", "Internal server error: " + e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    @PostMapping("/getModel")
    public ResponseEntity<?> getVehicleModel(@RequestBody String make){
        String substrigedMake = substringIncomingString(make);
        Integer makeId = vehicleService.getMakeIdByMake(substrigedMake);
        Map<String, String> errorResponse = new HashMap<>();
        try{
            List<String> modelList = vehicleService.getModelByMakeId(makeId);
            return ResponseEntity.ok(modelList);
        }catch (DataAccessException e){
            System.out.println(e.getMessage());
            errorResponse.put("message", "couldn't retrieve model"+ e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    @PostMapping("/getGeneration")
    public ResponseEntity<?> getVehicleGeneration(@RequestBody String model){
        String substrigedModel = substringIncomingString(model);
        Map<String, String> errorResponse = new HashMap<>();
        try{
            List<String> generationList = vehicleService.getGenerationByModelId(substrigedModel);
            return ResponseEntity.ok(generationList);
        }catch (DataAccessException e){
            System.out.println(e.getMessage());
            errorResponse.put("error", "couldn't retrieve generation"+ e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    @PostMapping("/getEngine")
    public ResponseEntity<?> getVehicleEngine(@RequestBody EngineModel engineModel){
        String substringModel = engineModel.getModel();
        String substringGeneration = engineModel.getGeneration();
        Map<String, String> errorResponse = new HashMap<>();
        try{
            List<String> engineList = vehicleService.getEngineByModelAndGeneration(substringModel, substringGeneration);
            return ResponseEntity.ok(engineList);
        }catch (DataAccessException e){
            System.out.println(e.getMessage());
            errorResponse.put("message", "couldn't retrieve engine"+ e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse);
        }
    }

    @GetMapping("/getPartGroup")
    private ResponseEntity<?> getPartGroup(){
        Map<String,String> response = new HashMap<>();
        try{
            List<PartGroup> list = new ArrayList<>();
            List<CarPartGroup> groupList= vehicleService.getPartsGroup();
            for(CarPartGroup pg : groupList){
                String base64Image = Base64.getEncoder().encodeToString(pg.getImage());
                String groupName = pg.getGroup_name();
                List<String> partsList = vehicleService.getFourCarPartsByGroupId(pg.getGroup_id());
                list.add( new PartGroup(groupName,base64Image, partsList));
            }
            return ResponseEntity.ok(list);
        }catch (DataAccessException e){
            System.out.println("message: "+e.getMessage());
            response.put("message", "error: "+e.getMessage());
             return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }


    @PostMapping("/getPartsList")
    private ResponseEntity<?> getPartsList(@RequestBody String partGroup){
        Map<String,String> response = new HashMap<>();
        String partGroupSub = substringIncomingString(partGroup);
        try{
            int groupId = vehicleService.getGroupIdByGroupName(partGroupSub);
            List<Part> list = new ArrayList<>();
            List<CarPartType> partsList = vehicleService.getPartsTypeByGroupId(groupId);
            for(CarPartType pg : partsList){
                String base64Image = Base64.getEncoder().encodeToString(pg.getImage());
                String partTypeName = pg.getPartTypeName();
                list.add(new Part(partTypeName,base64Image));
            }
            return ResponseEntity.ok(list);
        }catch (DataAccessException e){
            System.out.println("message: "+e.getMessage());
            response.put("message", "error: "+e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    @GetMapping("/getModelAll")
    public ResponseEntity<?> getEveryModel(){
        try {
            return ResponseEntity.ok(vehicleService.getEveryModel());
        }catch (Exception e){
            System.out.println(e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("error loading model " +e.getMessage());
        }
    }

    @GetMapping("/getGenerationAll")
    public ResponseEntity<?> getEveryGeneration(){
        try {
            List<Generation> list = vehicleService.getEveryGeneration();
            for (Generation o : list) {
                System.out.println(o.getGeneration()+" "+o.getGeneration_id());
            }
            return ResponseEntity.ok(vehicleService.getEveryGeneration());
        }catch (Exception e){
            System.out.println(e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("error loading generation " +e.getMessage());
        }
    }

    @GetMapping("/getEngineAll")
    public ResponseEntity<?> getEveryEngine(){
        try {
            return ResponseEntity.ok(vehicleService.getEveryEngine());
        }catch (Exception e){
            System.out.println(e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("error loading engine " +e.getMessage());
        }
    }

}

File: node-backend/controllers/VehicleController.js
const { buffer } = require('stream/consumers');
const vehicleModel = require('../models/VehicleModel');
const { response } = require('express');

async function getVehicleMake(req, res){
    try{
        const makeList = await vehicleModel.getVehicleMake();
        return res.status(200).json(makeList);
    }catch(err){
        console.log("Error in getVehicleMake");
        return res.status(500).json({error: err.message});
    }
}

async function getVehicleModel(req, res){
    try{
        let make = req.body.make;
        
        const makeId = await vehicleModel.getMakeIdByMake(make);
        const modelList = await vehicleModel.getModelByMakeId(makeId);
        
        const modelNames = modelList.map(obj => obj.model);
        return res.status(200).json(modelNames);
    }catch(err){
        console.log("Error in getVehicleModel");
        return res.status(500).json({error: err.message});
    }
}

async function getVehicleGeneration(req, res){
    try{
        const model = req.body.model;
        const generationList = await vehicleModel.getVehicleGenerationByModel(model);
        const generationNames = generationList.map(obj => obj.generation);
        return res.status(200).json(generationNames);
    }catch(err){
        console.log("Error in getVehicleGeneration");
        return res.status(500).json({error: err.message});
    }
}

async function getVehicleEngine(req, res){
    try{
        const model = req.body.model;
        const generation = req.body.generation;

        const modelId = await vehicleModel.getModelIdByModel(model);
        const generationId = await vehicleModel.getGenerationIdByGeneration(generation);

        const engineList = await vehicleModel.getVehicleEngineByGenerationAndModel(modelId, generationId);
        const engineNames = engineList.map(obj => obj.engine);

        return res.status(200).json(engineNames)
    }catch(err){
        console.log("Error in getVehicleEngine", err.message);
        return res.status(500).json({error: err.message});
    }
}

async function getPartGroup(req, res) {
  try {
    const loadedGroup = await vehicleModel.getPartsGroup();
    let groupList = [];

    const batchSize = 5;

    for(let i = 0; i< loadedGroup.length; i+= batchSize){
        const batch = loadedGroup.slice(i, i+batchSize);

        const results = await Promise.all(
            batch.map(async (element) => {
                const groupId = element.group_id;
                const image = element.image ? element.image.toString("base64") : null;
                let finalPartsList = [];
                const partsList = await vehicleModel.getFourCarPartsByGroupId(groupId);
                partsList.map(obj => finalPartsList.push(obj.part_type_name));
                return{
                    groupName: element.group_name,
                    groupImage: image,
                    partsList: finalPartsList
                };
            })
        )
        groupList.push(...results);
    }

    return res.status(200).json(groupList);
  } catch (err) {
    console.error("Error in getPartGroup", err);
    return res.status(500).json({ error: err.message });
  }
}

async function getPartsList(req, res){
    try{
        const groupName = req.body.groupName;
        const groupId = await vehicleModel.getGroupIdByGroupName(groupName);
        console.log(groupId);
        const partsList = await vehicleModel.getPartsTypeByGroupId(groupId);
        let finalPartsList = [];

        for(let i = 0; i< partsList.length; i++){
            const element = partsList[i];
            const image = element.image ? element.image.toString("base64") : null;
            
            const partTypeName = element.part_type_name;
            finalPartsList.push({
                partName: partTypeName,
                groupImage: image
            })
        }

        return res.status(200).json(finalPartsList);
    }catch(err){
        console.log("Error in getPartsList");
        return res.status(500).json({error: err.message});
    }
}

module.exports = {
    getVehicleMake,
    getVehicleModel,
    getVehicleGeneration,
    getVehicleEngine,
    getPartGroup,
    getPartsList
}